File: 1_insertion_sort/C/insertion_sort.c
#include <stdio.h>

void insertion_sort(int* A, int size) {
    int i, j, key;

    for (j = 1; j < size; j++) {
        i = j - 1;
        key = A[j];

        while (i >= 0 && key < A[i]) {
            A[i + 1] = A[i];
            i = i - 1;
        }
        A[i + 1] = key;
  }
}


int main() {   
    int i;
    int size = 6;
    int A[6] = { 5, 2, 4, 6, 1, 3 };

    insertion_sort(A, size);

    for (i = 0; i < size; i++) {
        printf ("%d ", A[i]);
    }    
}

File: 1_insertion_sort/java/InsertionSort.java
class InsertionSort implements Sorter {
    private int[] A;

    public InsertionSort(int[] A) {
        this.A = A;
    }

    public int[] sort() {
        for (int j = 1; j < A.length; j++) {
            int i = j - 1;
            int key = A[j];

            while (i >= 0 && key < A[i]) {
                A[i + 1] = A[i--];
            }
            A[i + 1] = key;
        }

        return A;
    }
}
